functions/modules/stats/index.js

```diff
@@ -116,120 +117,126 @@ exports.actionUpdate = functions
             addCO2eToStats(
                 newValues.uid,
                 newValues.category,
                 co2e,
                 actionCount,
-                newValues.creation_time,
-            )
+                newValues.creation_time
+            );
         }
 
         // Action is UPDATED
         // If action is updated, we substract the previous co2e and add the new co2e to the stats
         else {
-            const co2e = newValues.co2e - previousValues.co2e
-            const actionCount = 0
+            const co2e = newValues.co2e - previousValues.co2e;
+            const actionCount = 0;
             addCO2eToStats(
                 newValues.uid,
                 newValues.category,
                 co2e,
                 actionCount,
-                newValues.creation_time,
-            )
+                newValues.creation_time
+            );
         }
-    })
+    });
 
 exports.userUpdate = functions
     .region('europe-west6')
     .runWith({ minInstances: 1 })
     .firestore.document('/users/{documentId}')
     .onUpdate(async (event) => {
-        const previousValues = event.before.data()
-        const newValues = event.after.data()
+        const db = await dbInstance();
+
+        const previousValues = event.before.data();
+        const newValues = event.after.data();
+
+        const updates = {};
 
-        // Update stats
-        await db
+        // If user target is updated, increment eventUpdateTargetCount
+        if (previousValues.target != newValues.target) {
+            updates.eventUpdateTargetCount = FieldValue.increment(1);
+        }
+
+        // If user team is updated, increment eventUpdateTargetCount
+        if (previousValues.team != newValues.team) {
+            updates.eventUpdateTeamCount = FieldValue.increment(1);
+        }
+
+        // Perform update
+        const statsRef = await db
             .collection('stats')
             .where('uid', '==', newValues.uid)
             .limit(1)
-            .get()
-            .then((query) => {
-                const userStats = query.docs[0]
-
-                // If user target is updated, increment eventUpdateTargetCount
-                if (previousValues.target != newValues.target) {
-                    userStats.ref.update({
-                        eventUpdateTargetCount: FieldValue.increment(1),
-                    })
-                }
-
-                // If user team is updated, increment eventUpdateTargetCount
-                if (previousValues.team != newValues.team) {
-                    userStats.ref.update({
-                        eventUpdateTeamCount: FieldValue.increment(1),
-                    })
-                }
-            })
-    })
+            .get();
+        if (!statsRef.empty) {
+            await statsRef.docs[0].ref.update(updates);
+        }
+
+        // If user sponsor is updated, increment sponsorCount of sponsoring User
+        if (previousValues.sponsor != newValues.sponsor) {
```

Comment:
Need to check if the previous value is null / empty. Logically, we can't change our sponsor : we can't remove badge or decrement the sponsorCount, so it should be impossible to change